File: core/internal/payments/stripe.go
package payments

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"

	"github.com/stripe/stripe-go/v76"
	"github.com/stripe/stripe-go/v76/checkout/session"
	"github.com/stripe/stripe-go/v76/webhook"
)

func init() {
	// Set Stripe API key from environment
	stripe.Key = os.Getenv("STRIPE_SECRET_KEY")
}

// Product represents a purchasable item
type Product struct {
	PriceID string `json:"priceId"`
	Name    string `json:"name"`
	Price   int64  `json:"price"` // in cents
}

// Products available for purchase
var Products = map[string]Product{
	"battle_pass":  {PriceID: "price_battlepass_499", Name: "Battle Pass", Price: 499},
	"vip_monthly":  {PriceID: "price_vip_monthly_399", Name: "MazeChase+ Monthly", Price: 399},
	"starter_pack": {PriceID: "price_starter_499", Name: "Starter Pack", Price: 499},
	"coins_500":    {PriceID: "price_coins_099", Name: "500 Coins", Price: 99},
	"coins_2500":   {PriceID: "price_coins_399", Name: "2500 Coins", Price: 399},
	"coins_10000":  {PriceID: "price_coins_999", Name: "10000 Coins", Price: 999},
}

// CreateCheckoutSessionRequest is the request body for creating a checkout session
type CreateCheckoutSessionRequest struct {
	PriceID     string `json:"priceId"`
	UserID      string `json:"userId"`
	ProductName string `json:"productName"`
	SuccessURL  string `json:"successUrl"`
	CancelURL   string `json:"cancelUrl"`
}

// CreateCheckoutSession creates a Stripe checkout session
func CreateCheckoutSession(w http.ResponseWriter, r *http.Request) {
	var req CreateCheckoutSessionRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	params := &stripe.CheckoutSessionParams{
		PaymentMethodTypes: stripe.StringSlice([]string{"card"}),
		Mode:               stripe.String(string(stripe.CheckoutSessionModePayment)),
		SuccessURL:         stripe.String(req.SuccessURL),
		CancelURL:          stripe.String(req.CancelURL),
		ClientReferenceID:  stripe.String(req.UserID),
		LineItems: []*stripe.CheckoutSessionLineItemParams{
			{
				Price:    stripe.String(req.PriceID),
				Quantity: stripe.Int64(1),
			},
		},
	}

	s, err := session.New(params)
	if err != nil {
		json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
		return
	}

	json.NewEncoder(w).Encode(map[string]string{"sessionId": s.ID})
}

// HandleWebhook processes Stripe webhook events
func HandleWebhook(w http.ResponseWriter, r *http.Request) {
	const MaxBodyBytes = int64(65536)
	r.Body = http.MaxBytesReader(w, r.Body, MaxBodyBytes)

	payload := make([]byte, MaxBodyBytes)
	n, err := r.Body.Read(payload)
	if err != nil && err.Error() != "EOF" {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	payload = payload[:n]

	endpointSecret := os.Getenv("STRIPE_WEBHOOK_SECRET")
	event, err := webhook.ConstructEvent(payload, r.Header.Get("Stripe-Signature"), endpointSecret)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	switch event.Type {
	case "checkout.session.completed":
		var session stripe.CheckoutSession
		if err := json.Unmarshal(event.Data.Raw, &session); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		// Grant the purchase to the user
		handleSuccessfulPayment(session.ClientReferenceID, session.ID)

	case "customer.subscription.created":
		// Handle new subscription
		fmt.Println("New subscription created")

	case "customer.subscription.deleted":
		// Handle subscription cancellation
		fmt.Println("Subscription cancelled")
	}

	w.WriteHeader(http.StatusOK)
}

func handleSuccessfulPayment(userID, sessionID string) {
	// TODO: Update database to grant purchase
	// - Add coins to user balance
	// - Activate battle pass
	// - Enable VIP status
	fmt.Printf("✅ Payment successful for user %s (session: %s)\n", userID, sessionID)
}

// CheckSubscription checks if a user has an active subscription
func CheckSubscription(w http.ResponseWriter, r *http.Request) {
	// TODO: Query database for subscription status
	userID := r.URL.Query().Get("userId")
	
	// Placeholder - return false for now
	json.NewEncoder(w).Encode(map[string]interface{}{
		"userId": userID,
		"active": false,
	})
}

File: ui-web/src/lib/payments/stripe.ts
/**
 * Stripe Payment Integration
 * 
 * Real payment flow for MazeChase monetization
 */

import { loadStripe } from '@stripe/stripe-js';
import type { Stripe } from '@stripe/stripe-js';

// Initialize Stripe (use test key for now)
let stripePromise: Promise<Stripe | null>;

export function getStripe() {
  if (!stripePromise) {
    // Use test key - replace with live key in production
    const key = import.meta.env.PUBLIC_STRIPE_KEY || 'pk_test_YOUR_KEY_HERE';
    stripePromise = loadStripe(key);
  }
  return stripePromise;
}

// Product type
export type ProductKey = keyof typeof PRODUCTS;

// Product IDs (create these in Stripe Dashboard)
export const PRODUCTS = {
  BATTLE_PASS: {
    priceId: 'price_battlepass_499', // $4.99
    name: 'Battle Pass',
    price: 499, // cents
  },
  VIP_MONTHLY: {
    priceId: 'price_vip_monthly_399', // $3.99/month
    name: 'MazeChase+ Monthly',
    price: 399,
    recurring: true,
  },
  STARTER_PACK: {
    priceId: 'price_starter_499', // $4.99 (one-time)
    name: 'Starter Pack',
    price: 499,
  },
  COINS_500: {
    priceId: 'price_coins_099', // $0.99
    name: '500 Coins',
    price: 99,
  },
  COINS_2500: {
    priceId: 'price_coins_399', // $3.99
    name: '2500 Coins',
    price: 399,
  },
  COINS_10000: {
    priceId: 'price_coins_999', // $9.99
    name: '10000 Coins',
    price: 999,
  },
} as const;

export type ProductKey = keyof typeof PRODUCTS;

/**
 * Create a checkout session and redirect to Stripe
 */
export async function checkout(productKey: ProductKey, userId: string): Promise<void> {
  const product = PRODUCTS[productKey];
  
  try {
    // Call our backend to create checkout session
    const response = await fetch('/api/create-checkout-session', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        priceId: product.priceId,
        userId,
        productName: product.name,
        successUrl: `${window.location.origin}/payment/success?product=${productKey}`,
        cancelUrl: `${window.location.origin}/payment/cancel`,
      }),
    });

    const { sessionId, error } = await response.json();
    
    if (error) {
      throw new Error(error);
    }

    // Redirect to Stripe Checkout
    const stripe = await getStripe();
    if (stripe) {
      const { error: stripeError } = await stripe.redirectToCheckout({ sessionId });
      if (stripeError) {
        throw stripeError;
      }
    }
  } catch (err) {
    console.error('Payment error:', err);
    throw err;
  }
}

/**
 * Quick buy with Stripe Payment Links (no backend needed!)
 * Create these links in Stripe Dashboard
 */
export const PAYMENT_LINKS = {
  BATTLE_PASS: 'https://buy.stripe.com/test_battlepass',
  VIP_MONTHLY: 'https://buy.stripe.com/test_vip',
  STARTER_PACK: 'https://buy.stripe.com/test_starter',
} as const;

export function buyWithLink(product: keyof typeof PAYMENT_LINKS, userId: string) {
  const link = PAYMENT_LINKS[product];
  // Add user ID as client reference
  window.open(`${link}?client_reference_id=${userId}`, '_blank');
}

/**
 * Check if user has active subscription
 */
export async function checkSubscription(userId: string): Promise<boolean> {
  try {
    const response = await fetch(`/api/subscription/${userId}`);
    const { active } = await response.json();
    return active;
  } catch {
    return false;
  }
}

/**
 * Get user's purchase history
 */
export async function getPurchases(userId: string): Promise<string[]> {
  try {
    const response = await fetch(`/api/purchases/${userId}`);
    const { products } = await response.json();
    return products;
  } catch {
    return [];
  }
}
